### PR TITLE
Feature - Add canonical URL conversion support

### PR DESCRIPTION
**Pull request by lucatume**
Ticket: n/a

This PR contains the base code, tested in PR https://github.com/moderntribe/the-events-calendar/pull/2521

### CODE REVIEW

**Comment by lucatume** [comment]
Need to remove Context files from this PR.

File: src/Tribe/Rewrite.php [hunk 1]
```diff
@@ -295,4 +296,201 @@ protected function replace_array_key( &$array, $search, $replace ) {
...
+	}
+
+	/**
+	 * Returns an array of rewrite rules handled by the implementation.
+	 *
+	 * @since TBD
+	 *
+	 * @return array An array of rewrite rules handled by the implementation in the shape `[ <regex> => <path> ]`.
+	 */
+	protected function get_handled_rewrite_rules(  ) {
```
Comment (bordoni) [inline src/Tribe/Rewrite.php]: Spacing around `()`

File: src/Tribe/Rewrite.php [hunk 2]
```diff
@@ -295,4 +296,201 @@ protected function replace_array_key( &$array, $search, $replace ) {
...
+	}
+
+	/**
+	 * Returns a map relating localized regex matchers to query vars.
+	 *
+	 * @since TBD
+	 *
+	 * @return array A map of localized regex matchers in the shape `[ <localized_regex> => <query_var> ]`.
+	 */
+	protected function get_localized_matchers( ) {
```
Comment (bordoni) [inline src/Tribe/Rewrite.php]: Extra Spacing around `()`

File: src/Tribe/Rewrite.php [hunk 3]
```diff
@@ -295,4 +302,213 @@ protected function replace_array_key( &$array, $search, $replace ) {
...
+	 * Return a list of the query vars handled in the input rewrite rules.
+	 *
+	 * @since TBD
+	 *
+	 * @param array $rules A set of rewrite rules in the shape `[ <regex> => <path> ]`.
+	 *
+	 * @return array A list of all the query vars handled in the rules.
+	 */
+	protected function get_rules_query_vars( array $rules ) {
+		return array_unique( array_filter( array_merge( ...
```
Comment (bordoni) [inline src/Tribe/Rewrite.php]: What version `...` was included?

Comment (lucatume) [reply src/Tribe/Rewrite.php]: [5.6](http://docs.php.net/manual/en/migration56.new-features.php#migration56.new-features.splat)

Comment (sc0ttkclark) [reply src/Tribe/Rewrite.php]: whaaaat! Nice!

Variadic function parameters too? Awesome!

File: src/Tribe/Rewrite.php [hunk 4]
```diff
@@ -295,4 +302,232 @@ protected function replace_array_key( &$array, $search, $replace ) {
...
+	 * Returns an array of rewrite rules handled by the implementation.
+	 *
+	 * @since TBD
+	 *
+	 * @return array An array of rewrite rules handled by the implementation in the shape `[ <regex> => <path> ]`.
+	 */
+	protected function get_handled_rewrite_rules() {
+		global $wp_rewrite;
+		// While this is specific to The Events Calendar we're handling a small enough post type base to keep it here.
+		$pattern = '/post_type=tribe_(events|venue|organizer)/';
```
Comment (sc0ttkclark) [inline src/Tribe/Rewrite.php]: Should we have a filter for post types that we support and then pass those into a pattern?

**Pull request merged**